frb_codegen/src/generator/rust/ty_optional_list.rs

```diff
@@ -0,0 +1,52 @@
+use crate::generator::rust::ty::*;
+use crate::ir::*;
+use crate::target::Acc;
+use crate::target::Target;
+use crate::type_rust_generator_struct;
+
+use super::generate_import;
+use super::TypeGeneralListGenerator;
+
+type_rust_generator_struct!(TypeOptionalListGenerator, IrTypeOptionalList);
+
+impl TypeRustGeneratorTrait for TypeOptionalListGenerator<'_> {
+    fn wire2api_body(&self) -> Acc<Option<String>> {
+        Acc {
+            wasm: Some(TypeGeneralListGenerator::WIRE2API_BODY_WASM.to_owned()),
+            io: Some(TypeGeneralListGenerator::WIRE2API_BODY_IO.to_owned()),
+            ..Default::default()
+        }
+    }
+    fn wire_struct_fields(&self) -> Option<Vec<String>> {
+        Some(vec![
+            format!(
+                "ptr: *mut *mut {}",
+                self.ir.inner.rust_wire_type(Target::Io)
+            ),
+            "len: i32".to_string(),
+        ])
+    }
+    fn allocate_funcs(
+        &self,
+        collector: &mut super::ExternFuncCollector,
+        block_index: crate::utils::misc::BlockIndex,
+    ) -> Acc<Option<String>> {
+        Acc {
+            io: Some(collector.generate(
+                &format!("new_{}_{}", self.ir.safe_ident(), block_index),
+                [("len: i32", "int")],
+                Some(&format!("*mut {}", self.ir.rust_wire_type(Target::Io))),
+                &format!(
+                    "let wrap = {} {{ ptr: support::new_leak_vec_ptr(core::ptr::null_mut(), len), len }};
```

Comment:
nit: It seems that we are having some similarities with e.g. general_list. As usual, it it possible we somehow extract the common parts to reduce code duplication? (especially the "dangerous" part - unsafe leak ptr etc)

Reply:
I considered reusing generate_list_allocate_func, but I found it is not very appropriate for what OptionalList is trying to do and would require adjusting the function anyways.